core/src/main/java/com/sequenceiq/cloudbreak/reactor/handler/cluster/StopStartDownscaleGetRecoveryCandidatesHandler.java

```diff
@@ -0,0 +1,138 @@
+
+            StackDto stack = stackDtoService.getById(request.getResourceId());
+
+            Set<String> startedInstanceIdsOnCloudProvider = collectStartedInstancesFromCloudProvider(connector, ac, request.getAllInstancesInHostGroup())
+                    .stream()
+                    .map(CloudInstance::getInstanceId).collect(Collectors.toSet());
+
+            LOGGER.info("Started instanceIds on cloudProvider: {}", startedInstanceIdsOnCloudProvider);
+
+            flowMessageService.fireEventAndLog(stack.getId(), UPDATE_IN_PROGRESS.name(), CLUSTER_SCALING_STOPSTART_DOWNSCALE_IDENTIFYRECOVERYCANDIDATES,
+                    request.getHostGroupName());
+
+            List<InstanceMetadataView> startedInstanceMetadataWithServicesNotRunning =
+                    recoveryCandidateCollectionService.getStartedInstancesWithServicesNotRunning(stack, request.getHostGroupName(),
+                            startedInstanceIdsOnCloudProvider);
+
+            List<CloudInstance> instancesWithServicesNotRunning;
+
+            if (startedInstanceMetadataWithServicesNotRunning.isEmpty()) {
+                instancesWithServicesNotRunning = Collections.emptyList();
+            } else {
+                instancesWithServicesNotRunning = instanceMetaDataToCloudInstanceConverter.convert(startedInstanceMetadataWithServicesNotRunning,
+                        stack.getStack());
+                LOGGER.info("Collected {} downscale recovery candidates for hostGroup: {}", instancesWithServicesNotRunning.size(), request.getHostGroupName());
+            }
+
+            StopStartDownscaleGetRecoveryCandidatesResult result = new StopStartDownscaleGetRecoveryCandidatesResult(stack.getId(),
+                    instancesWithServicesNotRunning, request.getHostGroupName(), request.getHostIds());
+            notify(result, event);
+        } catch (Exception e) {
+            String errorMessage = String.format("Failed to collect downscale recovery candidates for stack: %s. " +
+                    "Proceeding with subsequent flow.", request.getResourceId());
+            LOGGER.warn(errorMessage, e);
+            StopStartDownscaleGetRecoveryCandidatesResult result = new StopStartDownscaleGetRecoveryCandidatesResult(request.getResourceId(),
+                    Collections.emptyList(), request.getHostGroupName(), request.getHostIds());
+            notify(result, event);
+        }
+    }
+
+    private List<CloudInstance> collectStartedInstancesFromCloudProvider(CloudConnector connector, AuthenticatedContext ac,
+            List<CloudInstance> cloudInstances) {
+        List<CloudVmInstanceStatus> vmInstanceStatuses = connector.instances().checkWithoutRetry(ac, cloudInstances);
+        return vmInstanceStatuses.stream()
+                .filter(vm -> InstanceStatus.STARTED.equals(vm.getStatus()))
+                .map(CloudVmInstanceStatus::getCloudInstance)
+                .collect(Collectors.toList());
+    }
```

Comment:
duplicates method, other handler has the same method

Reply:
I hadn't refactored this method because I wasn't sure if it would semantically fit into the common class. This method was added here because a subsequent state handler invokes the cloud-provider in a similar manner. Does this need to be refactored to the common `RecoveryCandidateCollectionService` class nonetheless?